helpers/aws.js

```diff
@@ -100,6 +104,89 @@ class AwsClient {
             });
         });
     }
+
+    /**
+     * Parse cache key
+     * @method getStoragePathForKey
+     * @param {String}              cacheKey          Path of the cache
+     * @return {String}
+     */
+    getStoragePathForKey(cacheKey) {
+        const convert = str => str
+            // Remove leading/trailing slashes
+            .replace(/(^\/|\/$)/g, '')
+            // Replace special URL characters
+            .replace(/[?&#%]/g, '~');
+
+        const parsedKey = convert(cacheKey);
+
+        return `${this.segment}/${parsedKey}`;
+    }
+
+    /**
+     * upload data as stream
+     * @method uploadAsStream
+     * @param {Object}              config               Config object
+     * @param {Stream}              config.payload       Payload to upload
+     * @param {String}              config.cacheKey      Path to cache
+     * @return {Promise}
+     */
+    uploadAsStream({ payload, cacheKey }) {
+        // stream the data to s3
+        const passthrough = new stream.PassThrough();
+        const params = {
+            Bucket: this.bucket,
+            Key: this.getStoragePathForKey(cacheKey),
+            Expires: new Date(),
+            ContentType: 'application/octet-stream',
+            Body: passthrough
+        };
+
+        payload.pipe(passthrough);
+
+        return this.client.upload(params).promise();
```

Comment:
Is it possible to pass stream to the existing catbox-s3 plugin? AFAIK it just delegates to the aws sdk.

Reply:
They don’t support it. None of the catbox plugins support interacting with stream data. Sadness, that’s why we need to do this extra logic
https://github.com/fhemberger/catbox-s3

Reply:
In fact, it is almost a straight pass through to aws sdk, for example, https://github.com/fhemberger/catbox-s3/blob/master/lib/index.js#L257, and `Body` can be a stream object.

I never try calling it but theoretically should behave the same, minus the concurrent upload. And since it's probably going to be on the deprecated list soon so im ok with either doing manual or use other simple s3 sugar npm package.

Reply:
Oh, i checked the doc, it seems like `pubObject` can take in `stream` as body now.  But from the doc it looks like `upload` is the recommended way for handling stream.
https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#putObject-property

Let me try with that to see if that works

Reply:
Tested `cache.set` directly with stream directly and it doesn't work... 
Got the err from the log: `data: Failed to stream the cache: Error: Cannot determine length of [object Object]`. Searched a little bit, it seems like for `pubObject`, it requires content-length for stream.

https://stackoverflow.com/questions/38442512/difference-between-upload-and-putobject-for-uploading-a-file-to-s3